### Importing the Libraries that are necessary to Train an Test the Data

In [164]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report
from sklearn.svm import SVC



### Loading the necessary glass data to apply the algorithm

In [165]:
data = pd.read_csv('glass.csv') #reading and loading the csv format glass data


### Seeing for the first five parameters to have a brief description regarding data

In [166]:
data.head()

RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Type
0  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0     1
1  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0     1
2  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0     1
3  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0     1
4  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0     1

In [167]:
data.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [168]:
print(data.isnull().sum())  # Check for null values
data.fillna(data.mean(), inplace=True)  # Replacing all the null values with the mean data for the particular column

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64


### Splitting the Data into training and testing parts with training data as 80% and 20% for Testing for the accuracy

In [169]:
X = data.drop('Type', axis=1)  # Extracting attributes, excluding 'Type' column
y = data['Type']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) #shuffling the data and using the splitting techique to segregate into train and test data


### Normalizing the Data

In [170]:
# Implement feature scaling using StandardScaler for data normalization.
# This step helps the algorithm perform better by standardizing feature scales.
from sklearn.preprocessing import StandardScaler
# Create a StandardScaler instance for feature scaling.
scaler = StandardScaler()
# Normalize the training data.
X_train = scaler.fit_transform(X_train)
# Normalize the testing data using the same scaler.
X_test = scaler.transform(X_test)
# End of feature scaling process.

### Applying the classifier Algorithm Linear Support Vector Machine

In [171]:
# Set up the SVM classifier with a linear kernel and a regularization parameter C of 5.
svm_class = SVC(kernel='linear', C=5) 
# Train the SVM classifier on the training data.
svm_class.fit(X_train, y_train)


SVC(C=5, kernel='linear')

### predicting for the Classifier

In [172]:
y_pred = svm_classifier.predict(X_test)


### Finding the accuracy percent how correctly it predicts

In [173]:
acc = round(svm_class.score(X_test, y_test) * 100, 2)
print(f"Accuracy of the Linear SVM model for the data after applying the algorith: {acc:.2f}")


Accuracy of the Linear SVM model for the data after applying the algorith: 76.74


### Looking for various features of classification report generated like precision, f1-score and seeing the accuracies

In [174]:
report = classification_report(y_test, y_pred, zero_division=1)

print("Classification Report for the data:\n", report)


Classification Report for the data:
               precision    recall  f1-score   support

           1       0.69      0.82      0.75        11
           2       0.69      0.79      0.73        14
           3       1.00      0.00      0.00         3
           5       0.80      1.00      0.89         4
           6       1.00      0.67      0.80         3
           7       1.00      0.88      0.93         8

    accuracy                           0.77        43
   macro avg       0.86      0.69      0.68        43
weighted avg       0.80      0.77      0.74        43



## We can conclude that the highest accuracy found is 77% accuracy for the test data and the accuracy is higher for SVM than Naive Bayes and the reasons are mentioned in the pdf attached below